File: apps/android/Components/NostrContent/NostrQuoteCard.kt
// Requires: compose-ui, compose-foundation, compose-material3,
// androidx.compose.material:material-icons-extended (for ChevronRight,
// FormatQuote, WarningAmber), io.coil-kt:coil-compose (>= 2.x). Kotlin 1.9+.
//
// Compose mirror of the SwiftUI `NostrQuoteCard`. Reusable quote / embedded-
// event card with four stable rendering contracts:
//
//   - .Collapsed → single-line "View quote" affordance; expands on tap
//   - .Compact   → author + truncated content text, border only
//   - .Rich      → avatar + name + truncated content + optional media thumb
//   - .Missing   → "Content unavailable" placeholder with unresolvedUri
//
// The app is responsible for hydrating the model; the card only renders.
// Depends on `compose/content-core`.

package nmp.content

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ChevronRight
import androidx.compose.material.icons.filled.FormatQuote
import androidx.compose.material.icons.filled.WarningAmber
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil.compose.SubcomposeAsyncImage

public enum class NostrQuoteCardVariant {
    /** Single-line `View quote` affordance; expands to `.Rich` on tap when wired. */
    Collapsed,

    /** Author + truncated content text, framed only by a border. */
    Compact,

    /** Author avatar + name + truncated content + optional media thumb. */
    Rich,

    /** "Content unavailable" placeholder; surfaces `unresolvedUri`. */
    Missing,
}

public data class NostrQuoteCardModel(
    val id: String,
    val unresolvedUri: String? = null,
    val authorPubkey: String? = null,
    val authorDisplayName: String? = null,
    val authorAvatarUrl: String? = null,
    val content: String = "",
    val mediaThumbnailUrl: String? = null,
    val createdAtDisplay: String? = null,
) {
    public companion object {
        public val Missing: NostrQuoteCardModel = NostrQuoteCardModel(id = "")
    }
}

@Composable
public fun NostrQuoteCard(
    model: NostrQuoteCardModel,
    modifier: Modifier = Modifier,
    variant: NostrQuoteCardVariant = NostrQuoteCardVariant.Rich,
    onTap: (() -> Unit)? = null,
    onExpand: (() -> Unit)? = null,
) {
    val renderer = LocalNostrContentRenderer.current
    val defaultTap: () -> Unit = { renderer.callbacks.onEventRefTap(model.id) }

    when (variant) {
        NostrQuoteCardVariant.Collapsed -> CollapsedBody(
            modifier = modifier,
            onTap = onExpand ?: onTap ?: defaultTap,
        )
        NostrQuoteCardVariant.Compact -> CompactBody(
            model = model,
            modifier = modifier,
            onTap = onTap ?: defaultTap,
        )
        NostrQuoteCardVariant.Rich -> RichBody(
            model = model,
            modifier = modifier,
            onTap = onTap ?: defaultTap,
        )
        NostrQuoteCardVariant.Missing -> MissingBody(
            model = model,
            modifier = modifier,
            onTap = onTap ?: defaultTap,
        )
    }
}

@Composable
private fun CollapsedBody(modifier: Modifier, onTap: () -> Unit) {
    val renderer = LocalNostrContentRenderer.current
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(8.dp))
            .background(renderer.quoteBackgroundColor)
            .border(1.5.dp, renderer.quoteBorderColor, RoundedCornerShape(8.dp))
            .clickable { onTap() }
            .padding(10.dp)
            .semantics { contentDescription = "View quoted post" },
    ) {
        Icon(
            imageVector = Icons.Filled.FormatQuote,
            contentDescription = null,
            tint = renderer.linkColor,
        )
        Text(
            text = "View quote",
            color = renderer.linkColor,
            fontWeight = FontWeight.SemiBold,
        )
    }
}

@Composable
private fun CompactBody(
    model: NostrQuoteCardModel,
    modifier: Modifier,
    onTap: () -> Unit,
) {
    QuoteCardFrame(modifier = modifier, onTap = onTap) {
        val renderer = LocalNostrContentRenderer.current
        Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
            Text(
                text = "@${authorLabel(model)}",
                color = renderer.textColor,
                fontWeight = FontWeight.SemiBold,
                fontSize = 12.sp,
            )
            Text(
                text = truncated(model.content, limit = 140),
                color = renderer.textColor,
                maxLines = 3,
                overflow = TextOverflow.Ellipsis,
            )
        }
    }
}

@Composable
private fun RichBody(
    model: NostrQuoteCardModel,
    modifier: Modifier,
    onTap: () -> Unit,
) {
    val renderer = LocalNostrContentRenderer.current
    QuoteCardFrame(modifier = modifier, onTap = onTap) {
        Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                QuoteAvatar(model = model, size = 26.dp)
                Column(modifier = Modifier.weight(1f)) {
                    Text(
                        text = "@${authorLabel(model)}",
                        color = renderer.textColor,
                        fontWeight = FontWeight.SemiBold,
                        fontSize = 12.sp,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )
                    val createdAt = model.createdAtDisplay
                    if (createdAt != null) {
                        Text(
                            text = createdAt,
                            color = renderer.secondaryTextColor,
                            fontFamily = FontFamily.Monospace,
                            fontSize = 11.sp,
                            maxLines = 1,
                        )
                    }
                }
                Icon(
                    imageVector = Icons.Filled.ChevronRight,
                    contentDescription = null,
                    tint = renderer.secondaryTextColor,
                )
            }
            if (model.content.isNotEmpty()) {
                Text(
                    text = truncated(model.content, limit = 240),
                    color = renderer.textColor,
                    maxLines = 6,
                    overflow = TextOverflow.Ellipsis,
                )
            }
            val thumb = model.mediaThumbnailUrl
            if (thumb != null) {
                QuoteThumbnail(url = thumb)
            }
        }
    }
}

@Composable
private fun MissingBody(
    model: NostrQuoteCardModel,
    modifier: Modifier,
    onTap: () -> Unit,
) {
    val renderer = LocalNostrContentRenderer.current
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(8.dp))
            .background(renderer.quoteBackgroundColor)
            .border(1.5.dp, renderer.quoteBorderColor, RoundedCornerShape(8.dp))
            .clickable { onTap() }
            .semantics { contentDescription = "Open quoted post" }
            .padding(10.dp),
    ) {
        Icon(
            imageVector = Icons.Filled.WarningAmber,
            contentDescription = null,
            tint = renderer.placeholderColor,
        )
        Column(
            verticalArrangement = Arrangement.spacedBy(2.dp),
            modifier = Modifier.weight(1f),
        ) {
            Text(
                text = "Content unavailable",
                color = renderer.textColor,
                fontWeight = FontWeight.SemiBold,
                fontSize = 12.sp,
            )
            val uri = model.unresolvedUri
            if (!uri.isNullOrEmpty()) {
                // Compose has no built-in middle-ellipsis. Show a head+tail
                // window manually so the schema portion of the URI stays
                // visible alongside the trailing identifier.
                Text(
                    text = middleEllipsize(uri, maxLength = 48),
                    color = renderer.secondaryTextColor,
                    fontFamily = FontFamily.Monospace,
                    fontSize = 11.sp,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
            }
        }
    }
}

@Composable
private fun QuoteCardFrame(
    modifier: Modifier,
    onTap: () -> Unit,
    content: @Composable () -> Unit,
) {
    val renderer = LocalNostrContentRenderer.current
    Box(
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(8.dp))
            .background(renderer.quoteBackgroundColor)
            .border(1.5.dp, renderer.quoteBorderColor, RoundedCornerShape(8.dp))
            .clickable { onTap() }
            .padding(10.dp)
            .semantics { contentDescription = "Open quoted post" },
    ) {
        content()
    }
}

@Composable
private fun QuoteAvatar(model: NostrQuoteCardModel, size: Dp) {
    val identityKey = model.authorPubkey ?: model.id
    val avatarUrl = model.authorAvatarUrl
    if (avatarUrl.isNullOrEmpty()) {
        QuoteAvatarFallback(identityKey = identityKey, size = size)
        return
    }
    SubcomposeAsyncImage(
        model = avatarUrl,
        contentDescription = null,
        contentScale = ContentScale.Crop,
        loading = { QuoteAvatarFallback(identityKey = identityKey, size = size) },
        error = { QuoteAvatarFallback(identityKey = identityKey, size = size) },
        modifier = Modifier
            .size(size)
            .clip(CircleShape),
    )
}

@Composable
private fun QuoteAvatarFallback(identityKey: String, size: Dp) {
    Box(
        contentAlignment = Alignment.Center,
        modifier = Modifier
            .size(size)
            .clip(CircleShape)
            .background(NostrIdenticon.colorForPubkey(identityKey)),
    ) {
        Text(
            text = NostrIdenticon.initialsForPubkey(identityKey),
            color = Color.White,
            fontWeight = FontWeight.SemiBold,
            fontSize = 12.sp,
        )
    }
}

@Composable
private fun QuoteThumbnail(url: String) {
    val renderer = LocalNostrContentRenderer.current
    SubcomposeAsyncImage(
        model = url,
        contentDescription = null,
        contentScale = ContentScale.Crop,
        loading = {
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .heightIn(min = 60.dp, max = 120.dp)
                    .clip(RoundedCornerShape(6.dp))
                    .background(renderer.codeBackgroundColor),
            )
        },
        error = {
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .heightIn(min = 60.dp, max = 120.dp)
                    .clip(RoundedCornerShape(6.dp))
                    .background(renderer.codeBackgroundColor),
            )
        },
        modifier = Modifier
            .fillMaxWidth()
            .heightIn(max = 160.dp)
            .clip(RoundedCornerShape(6.dp)),
    )
}

private fun authorLabel(model: NostrQuoteCardModel): String {
    val display = model.authorDisplayName
    if (!display.isNullOrEmpty()) return display
    val pubkey = model.authorPubkey
    if (!pubkey.isNullOrEmpty()) return shortPubkeyLocal(pubkey)
    return shortPubkeyLocal(model.id)
}

private fun shortPubkeyLocal(value: String): String {
    if (value.length <= 12) return value
    return "${value.take(8)}…${value.takeLast(4)}"
}

private fun truncated(value: String, limit: Int): String {
    val trimmed = value.trim()
    if (trimmed.length <= limit) return trimmed
    return "${trimmed.take(limit)}…"
}

private fun middleEllipsize(value: String, maxLength: Int): String {
    if (value.length <= maxLength) return value
    val half = (maxLength - 1) / 2
    return "${value.take(half)}…${value.takeLast(half)}"
}

File: apps/ios/Components/NostrContent/NostrQuoteCard.swift
import SwiftUI

/// Variant of the quote/embed card. Each style is a stable rendering contract
/// the embedding app can switch between without re-deriving data.
public enum NostrQuoteCardVariant: Equatable {
    /// Single-line `View quote` affordance; expands to `.rich` on tap when the
    /// app wires the `onExpand` closure.
    case collapsed
    /// Author + truncated content text, framed only by a border.
    case compact
    /// Author avatar stub + name + truncated content + optional media thumb.
    case rich
    /// "Content unavailable" placeholder, surfaces `unresolvedUri` to the user.
    case missing
}

/// Data model the embedding app populates per quote card. The card view is
/// pure — it never fetches network state. Apps pass a closure or environment
/// value into their data layer to hydrate `NostrQuoteCardModel` instances.
public struct NostrQuoteCardModel: Equatable, @unchecked Sendable {
    public var id: String
    public var unresolvedUri: String?
    public var authorPubkey: String?
    public var authorDisplayName: String?
    public var authorAvatarUrl: URL?
    public var content: String
    public var mediaThumbnailUrl: URL?
    public var createdAtDisplay: String?

    public init(
        id: String,
        unresolvedUri: String? = nil,
        authorPubkey: String? = nil,
        authorDisplayName: String? = nil,
        authorAvatarUrl: URL? = nil,
        content: String = "",
        mediaThumbnailUrl: URL? = nil,
        createdAtDisplay: String? = nil
    ) {
        self.id = id
        self.unresolvedUri = unresolvedUri
        self.authorPubkey = authorPubkey
        self.authorDisplayName = authorDisplayName
        self.authorAvatarUrl = authorAvatarUrl
        self.content = content
        self.mediaThumbnailUrl = mediaThumbnailUrl
        self.createdAtDisplay = createdAtDisplay
    }

    public static let missing = NostrQuoteCardModel(id: "")
}

/// Reusable quote / embedded-event card. The app is responsible for hydrating
/// the model; the card only renders.
public struct NostrQuoteCard: View {
    public var model: NostrQuoteCardModel
    public var variant: NostrQuoteCardVariant
    public var onTap: (() -> Void)?
    public var onExpand: (() -> Void)?

    @Environment(\.nostrContentRenderer) private var renderer

    public init(
        model: NostrQuoteCardModel,
        variant: NostrQuoteCardVariant = .rich,
        onTap: (() -> Void)? = nil,
        onExpand: (() -> Void)? = nil
    ) {
        self.model = model
        self.variant = variant
        self.onTap = onTap
        self.onExpand = onExpand
    }

    public var body: some View {
        switch variant {
        case .collapsed: collapsedBody
        case .compact:   compactBody
        case .rich:      richBody
        case .missing:   missingBody
        }
    }

    // MARK: variants

    private var collapsedBody: some View {
        Button {
            (onExpand ?? onTap ?? defaultTap)()
        } label: {
            HStack(spacing: 8) {
                Image(systemName: "quote.bubble")
                    .foregroundStyle(renderer.linkColor)
                Text("View quote")
                    .font(.callout.weight(.semibold))
                    .foregroundStyle(renderer.linkColor)
                Spacer(minLength: 0)
            }
            .padding(10)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(renderer.quoteBackgroundColor, in: RoundedRectangle(cornerRadius: 8))
            .overlay(border)
        }
        .buttonStyle(.plain)
        .accessibilityLabel("View quoted post")
    }

    private var compactBody: some View {
        cardButton {
            VStack(alignment: .leading, spacing: 4) {
                Text("@\(authorLabel)")
                    .font(.caption.weight(.semibold))
                    .foregroundStyle(renderer.textColor)
                Text(truncated(model.content, limit: 140))
                    .font(.callout)
                    .foregroundStyle(renderer.textColor)
                    .lineLimit(3)
            }
        }
    }

    private var richBody: some View {
        cardButton {
            VStack(alignment: .leading, spacing: 8) {
                HStack(spacing: 8) {
                    avatar
                    VStack(alignment: .leading, spacing: 1) {
                        Text("@\(authorLabel)")
                            .font(.caption.weight(.semibold))
                            .foregroundStyle(renderer.textColor)
                            .lineLimit(1)
                        if let display = model.createdAtDisplay {
                            Text(display)
                                .font(.caption2.monospaced())
                                .foregroundStyle(renderer.secondaryTextColor)
                                .lineLimit(1)
                        }
                    }
                    Spacer(minLength: 0)
                    Image(systemName: "chevron.right")
                        .font(.caption2.weight(.semibold))
                        .foregroundStyle(renderer.secondaryTextColor)
                }
                if !model.content.isEmpty {
                    Text(truncated(model.content, limit: 240))
                        .font(.callout)
                        .foregroundStyle(renderer.textColor)
                        .lineLimit(6)
                }
                if let thumb = model.mediaThumbnailUrl {
                    thumbnail(thumb)
                }
            }
        }
    }

    private var missingBody: some View {
        cardButton {
            HStack(spacing: 10) {
                Image(systemName: "exclamationmark.bubble")
                    .font(.callout.weight(.semibold))
                    .foregroundStyle(renderer.placeholderColor)
                VStack(alignment: .leading, spacing: 2) {
                    Text("Content unavailable")
                        .font(.caption.weight(.semibold))
                        .foregroundStyle(renderer.textColor)
                    if let uri = model.unresolvedUri, !uri.isEmpty {
                        Text(uri)
                            .font(.caption2.monospaced())
                            .foregroundStyle(renderer.secondaryTextColor)
                            .lineLimit(1)
                            .truncationMode(.middle)
                    }
                }
                Spacer(minLength: 0)
            }
        }
    }

    // MARK: helpers

    private func cardButton<Content: View>(@ViewBuilder content: () -> Content) -> some View {
        Button {
            (onTap ?? defaultTap)()
        } label: {
            content()
                .padding(10)
                .frame(maxWidth: .infinity, alignment: .leading)
                .background(renderer.quoteBackgroundColor, in: RoundedRectangle(cornerRadius: 8))
                .overlay(border)
        }
        .buttonStyle(.plain)
        .accessibilityLabel("Open quoted post")
    }

    private func defaultTap() {
        renderer.callbacks.onEventRefTap(model.id)
    }

    @ViewBuilder
    private var avatar: some View {
        if let avatarUrl = model.authorAvatarUrl {
            NostrImageView(url: avatarUrl)
                .frame(width: 26, height: 26)
                .clipShape(Circle())
        } else {
            avatarFallback
        }
    }

    private var avatarFallback: some View {
        NostrIdenticon.identiconView(
            forPubkey: model.authorPubkey ?? model.id,
            size: 26
        )
        .clipShape(Circle())
    }

    private func thumbnail(_ url: URL) -> some View {
        NostrImageView(url: url)
            .frame(maxWidth: .infinity, maxHeight: 160)
            .clipped()
            .clipShape(RoundedRectangle(cornerRadius: 6))
    }

    private var border: some View {
        RoundedRectangle(cornerRadius: 8)
            .stroke(renderer.quoteBorderColor, lineWidth: 1.5)
    }

    private var authorLabel: String {
        if let display = model.authorDisplayName, !display.isEmpty {
            return display
        }
        if let pubkey = model.authorPubkey, !pubkey.isEmpty {
            return shortPubkey(pubkey)
        }
        return shortPubkey(model.id)
    }

    private func truncated(_ value: String, limit: Int) -> String {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.count > limit else { return trimmed }
        return "\(trimmed.prefix(limit))…"
    }

    private func shortPubkey(_ value: String) -> String {
        guard value.count > 12 else { return value }
        return "\(value.prefix(8))…\(value.suffix(4))"
    }
}
